### PR TITLE
Refactor plugin manager

### PR DESCRIPTION
**Pull request by kke**
Refactored the plugin manager to be more robust and easier to extend in the future.

It can now install from file paths or URL's, for example:

```console
$ kontena plugin install https://github.com/kontena/kontena-plugin-upcloud/releases/download/v0.3.1/kontena-plugin-upcloud-0.3.1.gem
```

or 

```console
$ kontena plugin install /tmp/plugin.gem
```

### CODE REVIEW

**Comment by kke** [comment]
Conflicts resolved. 

IMHO this is more robust than #2420

File: cli/lib/kontena/cli/stacks/common.rb [hunk 1]
```diff
@@ -128,9 +128,10 @@ def generate_services(yaml_services)
       end
     end
 
-    def set_env_variables(stack, grid)
+    def set_env_variables(stack, grid, action)
       ENV['STACK'] = stack
       ENV['GRID'] = grid
+      ENV['ACTION'] = action
```
Comment (SpComb) [inline cli/lib/kontena/cli/stacks/common.rb]: This `action` is probably not part of this PR?

File: cli/lib/kontena/plugin_manager/installer.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+      # Upgrade an installed plugin
+      # @param plugin_name [String]
+      # @param pre [Boolean] upgrade to a prerelease version if available. Will happen always when the installed version is a prerelease version.
+      def upgrade
+        installed = installed(plugin_name)
+        pre = installed.version.prerelease?
+
+        raise "Plugin #{plugin_name} not installed" unless installed
+        latest = rubygems_client.latest_version(plugin_name, pre: pre)
```
Comment (SpComb) [inline cli/lib/kontena/plugin_manager/installer.rb]: This is missing the plugin prefix.... `kontena plugin install digitalocean` will try and upgrade to the `digitalocean` gem's most recent 1.2.0 version.

File: cli/lib/kontena/cli/plugins/install_command.rb [hunk 3]
```diff
@@ -1,29 +1,32 @@
...
+    def installer
+      Kontena::PluginManager::Installer.new(name, pre: pre?, version: version)
+    end
 
-      if installed_version
+    def execute
+      if installed?(name)
         installed = spinner "Upgrading plugin #{name.colorize(:cyan)}" do
-          Kontena::PluginManager.instance.upgrade_plugin(name, pre: pre?)
+          installer.upgrade
```
Comment (SpComb) [inline cli/lib/kontena/cli/plugins/install_command.rb]: This ignores the `-v` and `--pre` args when upgrading.

File: cli/lib/kontena/cli/plugins/uninstall_command.rb [hunk 4]
```diff
@@ -1,26 +1,22 @@
...
-      confirm unless forced?
-      spinner "Uninstalling plugin #{pastel.cyan(name)}" do |spin|
-        begin
-          Kontena::PluginManager.instance.uninstall_plugin(name)
-        rescue => ex
-          $stderr.puts pastel.red("#{ex.class.name} : #{ex.message}")
-          logger.error(ex)
-          spin.fail
-        end
+      exit_with_error "Plugin #{name} has not been installed" unless installed?(name)
```
Comment (SpComb) [inline cli/lib/kontena/cli/plugins/uninstall_command.rb]: Does this still have the issue of not allowing you to `kontena plugin uninstall` plugins that are broken with `LoadError`s? There's also an `installed?` check in the `Kontena::PluginManager::Uninstaller`

File: cli/lib/kontena/plugin_manager/common.rb [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+module Kontena
+  module PluginManager
+    module Common
+      Gem.autoload :DefaultUserInteraction, 'rubygems/user_interaction'
+      Gem.autoload :StreamUI, 'rubygems/user_interaction'
+
+      # Execute block without SafeYAML. Rubygems YAML readerdoes security internally.
+      def without_safe(&block)
+        SafeYAML::OPTIONS[:default_mode] = :unsafe if Object.const_defined?(:SafeYAML)
```
Comment (kke) [inline cli/lib/kontena/plugin_manager/common.rb]: Yes, this PR got outdated. Need to bring up to date.

File: cli/lib/kontena/cli/plugins/install_command.rb (hunk 3 above)
Comment (kke) [inline cli/lib/kontena/cli/plugins/install_command.rb]: I think the `installer` method above sets those.

Comment (SpComb) [inline cli/lib/kontena/cli/plugins/install_command.rb]: They work for the initial install, but upgrades for already installed plugins will always take the latest version... looks like `--pre` might actually work, but not the explicit `--version`.

File: cli/lib/kontena/plugin_manager/installer.rb (hunk 2 above)
Comment (kke) [inline cli/lib/kontena/plugin_manager/installer.rb]: Need to verify. Looks like RubygemsClient prefixes anything you pass to it.

File: cli/lib/kontena/cli/plugins/uninstall_command.rb (hunk 4 above)
Comment (kke) [inline cli/lib/kontena/cli/plugins/uninstall_command.rb]: Probably, I recall that was somehow fixed in the merged PR. Need to sync.

Comment (kke) [inline cli/lib/kontena/cli/plugins/uninstall_command.rb]: Yes, it used to just crash, but it was fixed by just warning if a plugin fails to load and continue loading, and it's in this PR too.

File: cli/lib/kontena/cli/plugins/install_command.rb (hunk 3 above)
Comment (kke) [inline cli/lib/kontena/cli/plugins/install_command.rb]: fixed in https://github.com/kontena/kontena/pull/2434/commits/ecaeae3bb1ae0f649b20daa222ff134b3e550907#diff-b7721e55d3ea3e85d3e6115f1cb42e44R42

File: cli/lib/kontena/cli/plugins/uninstall_command.rb (hunk 4 above)
Comment (SpComb) [inline cli/lib/kontena/cli/plugins/uninstall_command.rb]: I mean this case:

```
$ bin/kontena plugin uninstall digitalocean
 [error] Failed to load plugin: kontena-plugin-digitalocean from /home/kontena/.kontena/gems/2.3.1/gems/kontena-plugin-digitalocean-0.3.2
	Rerun the command with environment DEBUG=true set to get the full exception.
 [error] Plugin digitalocean has not been installed
```

Comment (SpComb) [inline cli/lib/kontena/cli/plugins/uninstall_command.rb]: This is a separate issue, as it also affects the current code.

Comment (kke) [inline cli/lib/kontena/cli/plugins/uninstall_command.rb]: It now uninstalls a broken plugin.

File: cli/lib/kontena/plugin_manager/loader.rb [hunk 6]
```diff
@@ -41,12 +41,13 @@ def load_plugin(path, spec)
         Kontena.logger.debug { "Loaded plugin #{spec.name}" } if plugin_debug?
 
         report_tracking
-        loaded_plugins << spec
         true
-      rescue LoadError, ScriptError => ex
+      rescue => ex
```
Comment (SpComb) [inline cli/lib/kontena/plugin_manager/loader.rb]: Need to keep the `LoadError, ScriptError`, because they are not of `StandardError`

```
$ bin/kontena plugin list
/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:120:in `require': cannot load such file -- droplet_kit (LoadError)
	from /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:120:in `require'
	from /home/kontena/.kontena/gems/2.3.1/gems/kontena-plugin-digitalocean-0.3.2/lib/kontena_cli_plugin.rb:4:in `<top (required)>'
	from /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:120:in `require'
	from /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:120:in `require'
	from /home/kontena/kontena/kontena/cli/lib/kontena/plugin_manager/loader.rb:40:in `load_plugin'
	from /home/kontena/kontena/kontena/cli/lib/kontena/plugin_manager/loader.rb:21:in `block in load_plugins'
	from /home/kontena/kontena/kontena/cli/lib/kontena/plugin_manager/loader.rb:17:in `each'
	from /home/kontena/kontena/kontena/cli/lib/kontena/plugin_manager/loader.rb:17:in `load_plugins'
	from /home/kontena/kontena/kontena/cli/lib/kontena/plugin_manager.rb:22:in `plugins'
	from /home/kontena/kontena/kontena/cli/lib/kontena/plugin_manager.rb:15:in `init'
```

**Comment by SpComb** [comment]
Better, now the broken plugins show up in `kontena plugin list`, and you can `uninstall` them:

```
$ bin/kontena plugin list
 [error] Failed to load plugin: kontena-plugin-digitalocean from /home/kontena/.kontena/gems/2.3.1/gems/kontena-plugin-digitalocean-0.3.2
	Rerun the command with environment DEBUG=true set to get the full exception.
NAME           VERSION     DESCRIPTION
app-command    0.1.0.rc1   Restores the "kontena app" subcommand back to Kontena CLI v1.4+
aws            0.3.0       Kontena Amazon Web Services plugin
digitalocean   0.3.2       Kontena DigitalOcean plugin
$ bin/kontena plugin uninstall digitalocean
 [error] Failed to load plugin: kontena-plugin-digitalocean from /home/kontena/.kontena/gems/2.3.1/gems/kontena-plugin-digitalocean-0.3.2
	Rerun the command with environment DEBUG=true set to get the full exception.
 [done] Uninstalling plugin digitalocean     
```

Ideally they would show up as broken, but this is better than before.

**Comment by kke** [comment]
image

**Comment by kke** [comment]
It can now install from file paths or urls, which would (at least partially) fix #2657

**Comment by SpComb** [comment]
Could you make a separate PR for the #2657 install-from-uri enhancement? I'd rather merge this refactoring first.

**Comment by kke** [comment]
Done
